### DISCUSSION TITLE
Système d'imports de bases publicodes basé sur NPM

### DISCUSSION THREAD

**Discussion by laem** [publicodes/publicodes#89] [Ideas]
Comme introduit dans #88 , quand la base de règles grossit on est confrontés au besoin de séparer la base de règles en plusieurs dépôts indépendants, potentiellement maintenus par différentes équipes.

Il semble tout à fait logique d'utiliser l'architecture déjà en place sur github `orga/dépot` + le système NPM. 

`datagir/nosgestesclimat` pourrait déléguer une sous-partie indépendante à `datagir/modèle-poisson` et ne référencer que les valeurs précalulées. 

Gros avantage : mieux segmenter les bases de code, et ne pas nécessiter de recalculer en permanence les résultats de la sous-partie tant qu'elle ne publie pas une nouvelle release sur NPM.

Il faudrait qu'on s'assure alors que l'utilisateur puisse explorer les calculs : 
- soit en chargeant à la demande (lazy loading) la base de règle de B sur le site A
- soit en pointant simplement vers B/documentation, ce qui suppose que chaque dépôt de règle déploie sa propre doc sur une URL fiable

Le sujet devient un peu plus compliqué quand on considère qu'il n'y a pas une valeur précalculée (comme notre sujet du menu poisson), mais plusieurs en fonction de paramètres. On aurait des variables "internes", non modifiables et pour lesquelles les valeurs par défaut seraient utilisées lors de ces appels, et des variables externes, exposées et donc variables. 

> On pourrait aussi imaginer que les paquets ne feraient aucun pré-calcul, mais juste assembler plusieurs bases gérées par différentes équipes.

Je laisse chacun parler de ses besoins. Le sujet de l'API est très proche, @johangirod @mquandalle on en a parlé de vive voix.

@maxgfr j'imagine que le sujet peut vous intéresser, notamment pour déléguer la responsabilité de certaines briques du mille feuille, au vu de 
https://github.com/SocialGouv/code-du-travail-numerique/tree/65328c3c6fdaa3d527b915f4fbd1d4c71f9eb0ea/packages/code-du-travail-modeles/src/modeles/conventions

**Comment by maxgfr** [comment]
Oh nice, le proposal !

**Reply by maxgfr** [reply]
@m-maillot et @carolineBda la discussion peut vous intéresser !

**Comment by mquandalle** [comment]
Je souscrit à l'idée de référencer des bases de publicodes externes, mais via des url HTTP plutôt qu'en se basant sur NPM :

Je me base essentiellement sur l'avis de Ryan Dahl, l'auteur de Node.js, qui a dit qu'intégrer NPM dans Node.js était une erreur. Cela créé une dépendance à NPM (pas trop grave dans notre cas) et c'est une source importante de complexité (gestion des versions, des registres privés, de l'authentification, etc.). Sur [Deno](https://deno.land/), on utilise des url pour référencer des modules externes.

L'utilisation d'url HTTP est compatible avec les paquets publiés sur NPM grâce à des services comme [unpkg](https://unpkg.com/), par exemple notre paquet modèle social est accessible en json ici : https://unpkg.com/modele-social

Cela permet aussi de référencer des publicodes privés en utilisant une URL github type https://raw.githubusercontent.com/SocialGouv/code-du-travail-numerique/master/packages/code-du-travail-modeles/src/modeles/conventions/automobiles.yaml avec un jeton d'authentification Github [[doc Deno](https://deno.land/manual@v1.8.0/linking_to_external_code/private)].

Le sujet de l'API est effectivement proche, puisqu'on peut référencer l'URL d'un CSV plutôt que d'un fichier publicodes.

**Reply by laem** [reply]
Excellent, merci pour le retour. Par contre en tant qu'hébergeur d'un paquet, sans passer par NPM, comment se passe la publication ? Il y a des helpers pour le faire ? Il faut déployer sur netfliy ?  Ou alors des services dédiés existent, comme unpkg qui fait ça ?

**Reply by laem** [reply]
@mquandalle vous réfléchissez à passer mon-entreprise sur Deno ?

**Reply by mquandalle** [reply]
Ça dépend de ce qu'on prend comme format en entrée. Si on accepte le yaml, on peut directement référencer les fichiers sources avec l'URL github https://raw.githubusercontent.com/datagir/nosgestesclimat/master/data/total.yaml (ça fonctionne comme ça avec Deno, on peut importer une fonction JS à partir d'une URL github).

Si on veut systématiquement un json transformé, alors il faut qu'il soit publié quelque part. Ça peut être dans un paquet NPM, auquel cas on peut le récupérer avec unpkg, ou alors sur un autre service (fichier [directement dans le dépôt Git](https://github.com/betagouv/taux-collectifs-cotisation-atmp), [artefact github action](https://docs.github.com/en/actions/advanced-guides/storing-workflow-data-as-artifacts), etc.)

Pas prévu de passer mon-entreprise sur Deno, c'est pas franchement mature pour faire des sites complets. En revanche pour faire une couche d'API HTTP au dessus de `publicodes/modele-social` ça me paraît un très bon outil.

**Reply by mquandalle** [reply]
Intéressant : Un exemple d'import d'URL pour un composant d'UI géré depuis framer : https://youtu.be/dMBYI7pTR4Q?t=323, on peut imaginer la même chose avec l'import d'une URL Google Sheet ou autre

**Comment by denismerigoux** [comment]
Je linke ici mon commentaire précédent sur un système de modules pour Publicode, parce que c'est lié : https://github.com/betagouv/publicodes/discussions/86#discussioncomment-1625146. Notamment sur la partie privée/publique de ce qu'il y a dans les modules. Il faut que vous vous mettiez d'accord sur la sémantique que vous voulez au niveau de Publicodes dans l'abstrait pour ensuite s'assurer que l’implémentation en JS y est bien fidèle.

**Comment by laem** [comment]
Peut-être que la release 44 et ses variables privées apportent un peu de grain à moudre à cette réflexion.

**Reply by johangirod** [reply]
Tout à fait : les règles privées ne seraient pas exposées à l'extérieur du module.

**Comment by johangirod** [comment]
En lien : https://github.com/betagouv/publicodes/discussions/183
